core/sdk-core/src/main/java/software/amazon/awssdk/core/SdkBytes.java

```diff
@@ -0,0 +1,159 @@
+ *
+ * This can be created via static methods, like {@link SdkBytes#fromByteArray(byte[])}. This can be converted to binary types
+ * via instance methods, like {@link SdkBytes#asByteArray()}.
+ */
+@SdkPublicApi
+public class SdkBytes {
+    final byte[] bytes;
+
+    /**
+     * @see #fromByteArray(byte[])
+     * @see #fromByteBuffer(ByteBuffer)
+     * @see #fromInputStream(InputStream)
+     * @see #fromUtf8String(String)
+     * @see #fromString(String, Charset)
+     */
+    SdkBytes(byte[] bytes) {
+        this.bytes = Validate.paramNotNull(bytes, "bytes");
+    }
+
+    /**
+     * Create {@link SdkBytes} from a Byte buffer. This will read the remaining contents of the byte buffer.
+     */
+    public static SdkBytes fromByteBuffer(ByteBuffer byteBuffer) {
+        Validate.paramNotNull(byteBuffer, "byteBuffer");
+        return new SdkBytes(BinaryUtils.copyBytesFrom(byteBuffer));
+    }
+
+    /**
+     * Create {@link SdkBytes} from a Byte array. This will copy the contents of the byte array.
+     */
+    public static SdkBytes fromByteArray(byte[] bytes) {
+        Validate.paramNotNull(bytes, "bytes");
+        return new SdkBytes(Arrays.copyOf(bytes, bytes.length));
+    }
+
+    /**
+     * Create {@link SdkBytes} from a string, using the provided charset.
+     */
+    public static SdkBytes fromString(String string, Charset charset) {
+        Validate.paramNotNull(string, "string");
+        Validate.paramNotNull(string, "charset");
+        return new SdkBytes(string.getBytes(charset));
+    }
+
+    /**
+     * Create {@link SdkBytes} from a string, using the UTF-8 charset.
+     */
+    public static SdkBytes fromUtf8String(String string) {
+        return fromString(string, StandardCharsets.UTF_8);
+    }
+
+    /**
+     * Create {@link SdkBytes} from an input stream. This will read all of the remaining contents of the stream, but will not
+     * close it.
+     */
+    public static SdkBytes fromInputStream(InputStream inputStream) {
+        Validate.paramNotNull(inputStream, "inputStream");
+        return new SdkBytes(invokeSafely(() -> IoUtils.toByteArray(inputStream)));
+    }
+
+    /**
+     * @return The output as a read-only byte buffer.
+     */
+    public final ByteBuffer asByteBuffer() {
+        return ByteBuffer.wrap(bytes).asReadOnlyBuffer();
+    }
+
+    /**
+     * @return A copy of the output as a byte array.
+     * @see #asByteBuffer() to prevent creating an additional array copy.
+     */
+    public final byte[] asByteArray() {
+        return Arrays.copyOf(bytes, bytes.length);
+    }
+
+    /**
+     * Retrieve the output as a string.
+     *
+     * @param charset The charset of the string.
+     * @return The output as a string.
+     */
+    public final String asString(Charset charset) {
+        return new String(bytes, charset);
+    }
+
+    /**
+     * @return The output as a utf-8 encoded string.
+     */
+    public final String asUtf8String() {
```

Comment:
worth documenting the exceptions this may throw if the binary data is not a utf8 string?

Reply:
Yes!